osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -35,147 +44,306 @@
+     * <p>
+     * The master keys (DLMS master or M-Bus Default) cannot be changed on a
+     * device, but can be generated for use in tests or with simulated devices.
+     *
+     * @return a new 128bits key, unencrypted.
+     */
+    public Map<SecurityKeyType, byte[]> generate128BitsKeysAndStoreAsNewKeys(String deviceIdentification,
+            List<SecurityKeyType> keyTypes) {
+        SecretTypes secretTypes = new SecretTypes();
+        GenerateAndStoreSecretsRequest request = this
+                .createGenerateAndStoreSecretsRequest(deviceIdentification, secretTypes);
+        secretTypes.getSecretType().addAll(keyTypes.stream().map(SecurityKeyType::toSecretType).collect(toList()));
+
+        GenerateAndStoreSecretsResponse response = this.secretManagementClient.generateAndStoreSecrets(request);
         TypedSecrets typedSecrets = response.getTypedSecrets();
         List<TypedSecret> typedSecretList = typedSecrets.getTypedSecret();
-
-        return typedSecretList.stream().filter(typedSecret -> typedSecret.getType() == secretType).findFirst();
+        this.validateGenerateAndStoreResponse(keyTypes, response, typedSecretList);
+        return this.convertSoapSecretsToSecretMapByType(typedSecrets.getTypedSecret());
     }
 
-    private GetSecretsRequest getSoapRequestForKey(String deviceIdentification, SecretType secretType) {
-        GetSecretsRequest request = new GetSecretsRequest();
+    private GenerateAndStoreSecretsRequest createGenerateAndStoreSecretsRequest(String deviceIdentification,
+            SecretTypes secretTypes) {
+        GenerateAndStoreSecretsRequest request = new GenerateAndStoreSecretsRequest();
         request.setDeviceId(deviceIdentification);
-        request.setSecretTypes(new SecretTypes());
-        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
-        secretTypeList.add(secretType);
+        request.setSecretTypes(secretTypes);
         return request;
     }
 
-    private byte[] decryptSoapSecret(String deviceIdentification, TypedSecret typedSecret) {
+    private void validateGenerateAndStoreResponse(List<SecurityKeyType> keyTypes,
+            GenerateAndStoreSecretsResponse response, List<TypedSecret> typedSecretList) {
+        this.validateOsgpResultAndTypedSecrets(response.getResult(), response.getTechnicalFault(),
+                response.getTypedSecrets(), keyTypes.size());
+        typedSecretList.forEach(ts -> {
+            if (ts.getSecret() == null) {
+                throw new IllegalStateException(String.format("Generated a NULL key (key_type=%s)", ts.getType()));
+            }
+        });
+    }
+
+    private void validateOsgpResultAndTypedSecrets(OsgpResultType result, Object fault, TypedSecrets typedSecrets,
+            int expectedNrKeys) {
+        if (!OsgpResultType.OK.equals(result)) {
+            throw new IllegalStateException("Could not generate and store keys in secret-mgmt: " + fault.toString());
```

Comment:
I am not sure whether fault could be `null` here or not, but as a general safety precaution I think it is almost always better (unless you would want the `NullPointerException`, which seems unlikely) to not include `.toString()` when you concatenate any object with a `String` as its String representation will be appended anyway if the object is not `null`.